src/rpc/blockchain.cpp

```diff
@@ -1570,6 +1571,309 @@ UniValue getchaintxstats(const JSONRPCRequest& request)
+    int64_t swtotal_size = 0;
+    int64_t swtotal_weight = 0;
+    int64_t utxo_size_inc = 0;
+    CAmount total_out = 0;
+    CAmount totalfee = 0;
+    CAmount minfee = MAX_MONEY;
+    CAmount maxfee = 0;
+    CAmount minfeerate = MAX_MONEY;
+    CAmount maxfeerate = 0;
+    std::vector<CAmount> fee_array;
+    std::vector<CAmount> feerate_array;
+    int64_t mintxsize = MAX_BLOCK_SERIALIZED_SIZE;
+    int64_t maxtxsize = 0;
+    std::vector<int64_t> txsize_array;
+
+    CBlock block;
+    ReadBlockCheckPruned(block, pindex);
+
+    for (const auto& tx : block.vtx) {
+        outputs += tx->vout.size();
+        CAmount tx_total_out = 0;
+        for (const CTxOut& out : tx->vout) {
+            utxo_size_inc += GetSerializeSize(out, SER_NETWORK, PROTOCOL_VERSION) + PER_UTXO_OVERHEAD;
+            tx_total_out += out.nValue;
+        }
+
+        if (tx->IsCoinBase()) {
+            continue;
+        }
+        total_out += tx_total_out;
+        inputs += tx->vin.size(); // Don't count coinbase's fake input
+        int64_t tx_size = tx->GetTotalSize();
+        txsize_array.push_back(tx_size);
+        total_size += tx_size;
+        mintxsize = std::min(mintxsize, tx_size);
+        maxtxsize = std::max(maxtxsize, tx_size);
+        int64_t weight = GetTransactionWeight(*tx);
+        total_weight += weight;
+
+        if (tx->HasWitness()) {
+            ++swtxs;
+            swtotal_size += tx_size;
+            swtotal_weight += weight;
+        }
+
+        CAmount tx_total_in = 0;
+        for (const CTxIn& in : tx->vin) {
+            CTransactionRef tx_in;
+            RpcGetTx(in.prevout.hash, tx_in);
+            CTxOut prevoutput = tx_in->vout[in.prevout.n];
+
+            tx_total_in += prevoutput.nValue;
+            utxo_size_inc -= GetSerializeSize(prevoutput, SER_NETWORK, PROTOCOL_VERSION) + PER_UTXO_OVERHEAD;
+        }
+        CAmount txfee = tx_total_in - tx_total_out;
+        assert(MoneyRange(txfee));
+        fee_array.push_back(txfee);
+        totalfee += txfee;
+        minfee = std::min(minfee, txfee);
+        maxfee = std::max(maxfee, txfee);
+
+        // New feerate uses satoshis per virtual byte instead of per serialized byte
+        CAmount feerate = CFeeRate(txfee, weight).GetTruncatedFee(WITNESS_SCALE_FACTOR);
+        feerate_array.push_back(feerate);
+
+        minfeerate = std::min(minfeerate, feerate);
+        maxfeerate = std::max(maxfeerate, feerate);
+    }
+
+    for (const std::string& stat : stats) {
+        // Update map_stats
+        if (stat == "height") {
+            map_stats[stat].push_back((int64_t)pindex->nHeight);
+        } else if (stat == "time") {
+            map_stats[stat].push_back(pindex->GetBlockTime());
+        } else if (stat == "mediantime") {
+            map_stats[stat].push_back(pindex->GetMedianTimePast());
+        } else if (stat == "subsidy") {
+            map_stats[stat].push_back(GetBlockSubsidy(pindex->nHeight, Params().GetConsensus()));
```

Comment:
I think the values, fees and fee rates should be wrapped with ValueFromAmount, so outputs are decimal bitcoins rather than integer satoshis, for consistency with other RPC calls.

Reply:
No, we want to allow the slow fetch too. If the data is calculable we want to provide it no matter if more slowly. This rpc call shouldn't be expected to be fast anyway.

Reply:
I'm presuming this reply should be to the "GetTransaction(..., false)" comment. Getting fee info without txindex seems like a weird thing to do to me, but if it's supported then presumably the getblockstats RPC help text doesn't need to still say "It won't work .. without -txindex." ?

Reply:
Yeah, sorry. Was meant for the other comment. Actually looking at the code again the slow search will only find txs in the utxo, so using false should be fine. I'll test it with false.

Regarding using BTC instead of satoshis, I think we want to move everything to satoshis but we don't do it because it would be too disruptive, so we only do it with new calls.
The caller can trivially divide by 100000000 to get BTC if they want t show that.